src/backend/mpi/coherence.cpp

```diff
@@ -0,0 +1,208 @@
+/**
+ * @file
+ * @brief This file implements selective coherence mechanisms for ArgoDSM
+ * @copyright Eta Scale AB. Licensed under the Eta Scale Open Source License. See the LICENSE file for details.
+ */
+
+#include "../backend.hpp"
+#include "swdsm.h"
+#include "write_buffer.hpp"
+#include "virtual_memory/virtual_memory.hpp"
+
+// EXTERNAL VARIABLES FROM BACKEND
+/**
+ * @brief This is needed to access page information from the cache
+ * @deprecated Should be replaced with a cache API
+ */
+extern control_data *cacheControl;
+/**
+ * @brief globalSharers is needed to access and modify the pyxis directory
+ * @deprecated Should eventually be handled by a cache module
+ */
+extern unsigned long *globalSharers;
+/**
+ * @brief A cache mutex protects all operations on cacheControl
+ * @deprecated Should eventually be handled by a cache module
+ */
+extern pthread_mutex_t cachemutex;
+/**
+ * @brief ibsem is used to serialize all Infiniband (MPI) operations
+ * @deprecated Should not be needed once the cache module is implemented
+ */
+extern sem_t ibsem;
+/**
+ * @brief barwindowsused is needed to know which globalDataWindows
+ * to close.
+ * @deprecated No longer needed once proper API functions are in place
+ */
+extern char * barwindowsused;
+/**
+ * @brief globalDataWindows need to be explicitly closed by the
+ * caller of storepageDIFF
+ * @deprecated Should eventually be handled through API functions
+ */
+extern MPI_Win *globalDataWindow;
+/**
+ * @brief sharerWindow protects the pyxis directory
+ * @deprecated Should not be needed once the pyxis directory is
+ * managed from elsewhere through a cache module.
+ */
+extern MPI_Win sharerWindow;
+/**
+ * @brief Needed to update argo statistics
+ * @deprecated Should be replaced by API calls to a stats module
+ */
+extern argo_statistics stats;
+/**
+ * @brief Needed to update information about cache pages touched
+ * @deprecated Should eventually be handled by a cache module
+ */
+extern argo_byte *touchedcache;
+/**
+ * @brief workcomm is needed to poke the MPI system during one sided RMA
+ */
+extern MPI_Comm workcomm;
+/**
+ * @brief Write buffer to ensure selectively handled pages can be removed
+ * @deprecated This should eventually be handled by a cache module
+ */
+extern write_buffer<std::size_t>* argo_write_buffer;
+
+/** @brief The size of an ArgoDSM page */
+const std::size_t pagesize = GRAN;
+
+namespace argo {
+	namespace backend {
+		void selective_acquire(void *addr, size_t size){
+			if(size == 0){
+				// Nothing to invalidate
+				return;
+			}
+
+			const std::size_t block_size = pagesize*CACHELINE;
+			const std::size_t start_address = reinterpret_cast<std::size_t>(argo::virtual_memory::start_address());
```

Comment:
This may be worth thinking about. I believe that, in concept, it makes sense for the default representation of an "address" return type to be `void*`.

However, we use the returned pointer from `start_address()` primarily in pointer arithmetics, in which case we have traditionally cast it to `char*`. It also is used directly to `mprotect` in a couple of places.
In this patch I have since our type discussion used `std::size_t` to represent addresses, hence the cast to `std::size_t` instead of `char*`.

Which would be the correct signature in this case? Or is it simply more _safe_ to return a `void*` and let each use cast as appropriate if necessary?